File: Dashboard/src/shared/channelGatewayScope.js
/**
 * Mirrors Sources/PluginSDK/ChannelGatewayScope.swift — Telegram topic-scoped channel ids.
 */

const UNIT_SEP = "\u001E";
const TOPIC_MARKER = "tgthread:";

const NEEDLE = `${UNIT_SEP}${TOPIC_MARKER}`;

/** Base gateway binding id (strip `\u001Etgthread:…` suffix when present). */
export function gatewayBindingChannelId(channelId) {
  const raw = String(channelId || "").trim();
  const idx = raw.indexOf(NEEDLE);
  if (idx === -1) {
    return raw;
  }
  const base = raw.slice(0, idx).trim();
  return base || raw;
}

/**
 * Whether an open session belongs to this project/binding channel (exact base or topic under that base).
 */
export function sessionChannelMatchesBinding(sessionChannelId, bindingChannelId) {
  const binding = String(bindingChannelId || "").trim();
  const session = String(sessionChannelId || "").trim();
  if (!binding) {
    return false;
  }
  if (session === binding) {
    return true;
  }
  return session.startsWith(`${binding}${NEEDLE}`);
}

File: Sources/PluginSDK/ChannelGatewayScope.swift
import Foundation

/// Scopes gateway traffic (e.g. Telegram forum topics) onto distinct Sloppy channel/session keys while preserving a stable binding id.
public enum ChannelGatewayScope {
    private static let unitSeparator = "\u{001E}" // RECORD SEPARATOR — unlikely in user channel ids
    private static let topicMarker = "tgthread:"

    /// Returns a distinct channel id per topic when `topicKey` is non-empty; otherwise `baseChannelId`.
    public static func scopedChannelId(baseChannelId: String, topicKey: String?) -> String {
        let base = baseChannelId.trimmingCharacters(in: .whitespacesAndNewlines)
        let topic = topicKey?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        guard !base.isEmpty, !topic.isEmpty else {
            return base
        }
        return "\(base)\(unitSeparator)\(topicMarker)\(topic)"
    }

    /// Splits a scoped id into binding base + optional topic key (Telegram `message_thread_id` string).
    public static func parse(_ scopedOrBaseChannelId: String) -> (baseChannelId: String, topicKey: String?) {
        let raw = scopedOrBaseChannelId.trimmingCharacters(in: .whitespacesAndNewlines)
        let needle = "\(unitSeparator)\(topicMarker)"
        guard let range = raw.range(of: needle) else {
            return (raw, nil)
        }
        let base = String(raw[..<range.lowerBound]).trimmingCharacters(in: .whitespacesAndNewlines)
        let topic = String(raw[range.upperBound...]).trimmingCharacters(in: .whitespacesAndNewlines)
        if base.isEmpty {
            return (raw, nil)
        }
        return (base, topic.isEmpty ? nil : topic)
    }

    /// Whether an open session was opened for this gateway binding (exact base or topic under that base).
    public static func sessionMatchesBinding(sessionChannelId: String, bindingChannelId: String) -> Bool {
        let binding = bindingChannelId.trimmingCharacters(in: .whitespacesAndNewlines)
        let session = sessionChannelId.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !binding.isEmpty else { return false }
        if session == binding { return true }
        return session.hasPrefix("\(binding)\(unitSeparator)\(topicMarker)")
    }
}
